File: 2021/day-02.py
def part1(commands):
    horizontal = 0
    depth = 0
    for (cmd, val) in commands:
        if cmd == "forward":
            horizontal += val
        elif cmd == "up":
            depth -= val
        elif cmd == "down":
            depth += val

    return horizontal * depth


def part2(commands):
    horizontal = 0
    depth = 0
    aim = 0
    for (cmd, val) in commands:
        if cmd == "forward":
            horizontal += val
            depth += aim * val
        elif cmd == "up":
            aim -= val
        elif cmd == "down":
            aim += val

    return horizontal * depth


with open("input-02.txt", "r") as f:
    commands = list(
        map(lambda item: (item[0], int(item[1])), [line.strip().split() for line in f])
    )

    print("Part 1:", part1(commands))
    print("Part 2:", part2(commands))

File: 2015/day-02.ts
import * as fs from "fs";
import * as path from "path";

const area = ([l, h, w]: number[]) => {
  return 2 * l * w + 2 * w * h + 2 * h * l;
};

const part1 = (input: number[][]) =>
  input
    .map(
      (item) =>
        area(item) +
        item
          .sort((a, b) => a - b)
          .slice(0, 2)
          .reduce((p, c) => p * c)
    )
    .reduce((p, c) => p + c, 0);

const part2 = (input: number[][]) =>
  input
    .map(
      (item) =>
        item
          .sort((a, b) => a - b)
          .slice(0, 2)
          .reduce((p, c) => p * 2 + c * 2) + item.reduce((p, c) => p * c)
    )
    .reduce((p, c) => p + c, 0);

fs.readFile(path.join(__dirname, "input-02.txt"), (err, data) => {
  if (err) throw err;
  let dims = data
    .toString()
    .split("\r\n")
    .map((size) => size.split("x").map((value) => Number(value)));

  console.log("Part 1: " + part1(dims));
  console.log("Part 2: " + part2(dims));
});
